src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -71,155 +58,270 @@ class TaskReplaceActor(
+          }
+        }.map(Killed).pipeTo(self)
+
+    case KillNext =>
+      // Pick first active old instance that has goal running
+      instances.valuesIterator.find { instance =>
+        instance.runSpecVersion < runSpec.version && instance.state.goal == Goal.Running
+      } match {
+        case Some(doomed) => killNextOldInstance(doomed).map(_ => Killed(Seq(doomed.instanceId))).pipeTo(self)
+        case None => self ! Killed(Seq.empty)
+      }
 
-  }
+    case Killed(killIds) =>
+      logger.info(s"Marking $killIds as stopped.")
+      // TODO(karsten): We may want to wait for `InstanceChanged(instanceId, ..., Goal.Stopped | Goal.Decommissioned)`.
+      // We mark the instance as doomed so that we won't select it in the next run.
+      killIds.foreach { instanceId =>
+        val killedInstance = instances(instanceId)
+        val updatedState = killedInstance.state.copy(goal = Goal.Stopped)
+        instances += instanceId -> killedInstance.copy(state = updatedState)
+      }
+
+      context.become(launching)
+      self ! ScheduleReadiness
 
-  override def instanceConditionChanged(instanceId: Instance.Id): Unit = {
-    if (healthyInstances(instanceId) && readyInstances(instanceId)) killNextOldInstance(Some(instanceId))
-    checkFinished()
+    // Stash all instance changed events
+    case stashMe: AnyRef =>
+      stash()
   }
 
-  def reconcileAlreadyStartedInstances(): Unit = {
-    logger.info(s"reconcile: found ${instancesAlreadyStarted.size} already started instances " +
-      s"and ${oldInstanceIds.size} old instances: ${currentInstances.map{ i => i.instanceId -> i.state.condition }}")
-    instancesAlreadyStarted.foreach(reconcileHealthAndReadinessCheck)
+  // Launch next new instance
+  def launching: Receive = {
+    case ScheduleReadiness =>
+      // Schedule readiness check for new healthy instance that has no active check yet.
+      instances.valuesIterator.find { instance =>
+        val noReadinessCheckScheduled = !instancesReady.contains(instance.instanceId)
+        instance.runSpecVersion == runSpec.version && instance.state.condition.isActive && instance.state.goal == Goal.Running && noReadinessCheckScheduled
+      } foreach { instance =>
+        initiateReadinessCheck(instance)
+
+        // Mark new instance as not ready
+        instancesReady += instance.instanceId -> false
+      }
+      self ! LaunchNext
+
+    case LaunchNext =>
+      logger.info("Launching next instance")
+      val oldActiveInstances = instances.valuesIterator.count { instance =>
+        instance.runSpecVersion < runSpec.version && instance.state.condition.isActive && instance.state.goal == Goal.Running
+      }
+      val newInstancesStarted = instances.valuesIterator.count { instance =>
+        instance.runSpecVersion == runSpec.version && instance.state.goal == Goal.Running
+      }
+      launchInstances(oldActiveInstances, newInstancesStarted).pipeTo(self)
+
+    case scheduledInstances: Seq[Instance] =>
```

Comment:
This would be cleaner if this actor would actually schedule the instances and not the launch queue.